x/campaign/keeper/msg_server_mint_vouchers.go

```diff
@@ -0,0 +1,60 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	spnerrors "github.com/tendermint/spn/pkg/errors"
+	"github.com/tendermint/spn/x/campaign/types"
+	profiletypes "github.com/tendermint/spn/x/profile/types"
+)
+
+func (k msgServer) MintVouchers(goCtx context.Context, msg *types.MsgMintVouchers) (*types.MsgMintVouchersResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	campaign, found := k.GetCampaign(ctx, msg.CampaignID)
+	if !found {
+		return nil, sdkerrors.Wrapf(types.ErrCampaignNotFound, "%d", msg.CampaignID)
+	}
+
+	// Get the coordinator ID
+	coordinatorID, found := k.profileKeeper.CoordinatorIDFromAddress(ctx, msg.Coordinator)
+	if !found {
+		return nil, sdkerrors.Wrap(profiletypes.ErrCoordAddressNotFound, msg.Coordinator)
+	}
+	if campaign.CoordinatorID != coordinatorID {
+		return nil, sdkerrors.Wrap(profiletypes.ErrCoordInvalid, fmt.Sprintf(
+			"coordinator of the campaign is %v",
+			campaign.CoordinatorID,
+		))
+	}
+
+	// Increase the campaign shares
+	campaign.AllocatedShares = types.IncreaseShares(campaign.AllocatedShares, msg.Shares)
+	if types.IsTotalSharesReached(campaign.AllocatedShares, campaign.TotalShares) {
+		return nil, sdkerrors.Wrapf(types.ErrTotalSharesLimit, "%v", msg.CampaignID)
+	}
+	k.SetCampaign(ctx, campaign)
```

Comment:
maybe it's better to update the campaign in the final of the method. If there is an error, the campaign was already updated, creating an invariant issue

Reply:
The context is only updated if the message is successful, but maybe it's better visually to put it at the end actually